cdb2api/cdb2api.c

```diff
@@ -1081,20 +1081,22 @@ struct cdb2_hndl {
     char stack[MAX_STACK];
     int send_stack;
     void *user_arg;
-    int gbl_event_version; /* Cached global event version */
+    int *gbl_event_version; /* Cached global event version */
     int api_call_timeout;
     int connect_timeout;
     int comdb2db_timeout;
     int socket_timeout;
     int request_fp; /* 1 if requesting the fingerprint; 0 otherwise. */
-    cdb2_event events;
+    cdb2_event *events;
```

Comment:
I don't think we need to inherit parent handle's events. It's much easier to just let the child handle get the global events. Doing a shallow copy here seems problematic too: would `refresh_gbl_events_on_hndl()` be called on both parent and child handle, resulting in an event being added twice?

Reply:
If we don't inherent parent handle's events then child handle will not get local events, is that ok?
Child and parent would both have the same events head pointer and version number. If an event is updated on the child handle it will also be updated on the parent handle since the head pointer always stays the same and is never updated

Reply:
When refresh gbl events is called on one of the handles, the events will be updated and the version number will be updated. Both handles have the same pointer to the events and version number. When refresh gbl events is called on the other handle, it will see that the event version numbers are already equal and will do nothing

Reply:
Okay makes sense. Can you move these 2 lines to `free_events()`

```
 free(tmp.events);
 free(tmp.gbl_event_version);
```

Reply:
I decided to put it outside bc if you put it inside and call free_events() followed by register_event() then you will get a seg fault bc hndl->events will be NULL once register_event is called. I guess free_events() is an internal function and this case never happens currently, but figured you should be able to add more events after free_events() is called?

Reply:
Or I can just add a comment to free_events() saying you can't add more events after it's called

Reply:
free_event() is invoked from cdb2_close(). I don't think you can operate on a handle after it's closed. You can still register events globally (by passing a NULL handle pointer).

Reply:
free_events() is also called in other functions like comdb2db_get_dbhosts which doesn't even call cdb2_open.  It calls cdb2_read_record which eventually populates the global events to the handle. And since cdb2_open is never called I guess you can't call cdb2_close on it and you need to manually free the events. So when you call free_events() outside of cdb2_close maybe you still might want to add more events later

Reply:
I can still just add my frees to that function tho and add a clarifying comment. Should I do that?

Reply:
We construct a partial cdb2_hndl struct in comdb2db_get_dbhosts(), in order to reuse some of the cdb2 routines to talk to comdb2db. That handle is on the stack.

Reply:
Yeah but the events are still on the heap, and since I'm moving the events head to the heap as well we might not want to free it until we know we're totally done with the handle

Reply:
Or I can also just calloc events head again in cdb2_register_event like I do in refresh_gbl_events_on_hndl. That might be better

Reply:
That's my point - you're pretty much done with the handle, when free_events() is called:

1) cdb2_close() - you can't use the handle after this point.
2) fake handle in comdb2db_get_dbhosts() - you can't use the handle after exiting from this function, as the handle is on the stack.

Reply:
I'll do this ^^

Reply:
Ok sounds good I'll add frees to function and also just add check in cdb2_register_event just in case.

Yeah I know in the code rn wherever free_event() is called it's in the same places where we are done with the handle. But I was thinking maybe in the future there could be places where free_event() could be called when you don't want to be done with handle

Reply:
Moved frees to function + added checks!

Reply:
https://github.com/bloomberg/comdb2/compare/5911788cbe6f9bd0a3dc275d3d0b20fad440815b..e669d41d68a6880ceaf92c5a5488370ff4b92874

Reply:
Ok done!
https://github.com/bloomberg/comdb2/compare/5911788cbe6f9bd0a3dc275d3d0b20fad440815b..e669d41d68a6880ceaf92c5a5488370ff4b92874
https://github.com/bloomberg/comdb2/compare/e669d41d68a6880ceaf92c5a5488370ff4b92874..9a125352fc578e85e5e10c50016e98f7c4fa7305